File: src/test/java/cube8540/oauth/authentication/users/domain/UserCredentialsKeyTest.java
package cube8540.oauth.authentication.users.domain;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import static cube8540.oauth.authentication.users.domain.UserTestHelper.EXPIRATION_DATETIME;
import static cube8540.oauth.authentication.users.domain.UserTestHelper.RAW_CREDENTIALS_KEY;
import static cube8540.oauth.authentication.users.domain.UserTestHelper.makeDefaultClock;
import static cube8540.oauth.authentication.users.domain.UserTestHelper.makeExpiredClock;
import static org.junit.jupiter.api.Assertions.assertEquals;

@DisplayName("인증키 테스트")
class UserCredentialsKeyTest {

    @Test
    @DisplayName("새로운 키 생성")
    void createNewKey() {
        UserCredentialsKey.setClock(makeDefaultClock());

        UserCredentialsKey key = new UserCredentialsKey(RAW_CREDENTIALS_KEY);
        assertEquals(EXPIRATION_DATETIME, key.getExpiryDateTime());
    }

    @Test
    @DisplayName("서로 다른 키 매칭")
    void matchedDifferentKey() {
        UserCredentialsKey.setClock(makeDefaultClock());
        UserCredentialsKey key = new UserCredentialsKey(RAW_CREDENTIALS_KEY);

        UserKeyMatchedResult result = key.matches("NOT MATCHED KEY");
        assertEquals(UserKeyMatchedResult.NOT_MATCHED, result);
    }

    @Test
    @DisplayName("만료된 키 매칭")
    void matchedExpiredKey() {
        UserCredentialsKey.setClock(makeDefaultClock());
        UserCredentialsKey key = new UserCredentialsKey(RAW_CREDENTIALS_KEY);

        UserCredentialsKey.setClock(makeExpiredClock());

        UserKeyMatchedResult result = key.matches(RAW_CREDENTIALS_KEY);
        assertEquals(UserKeyMatchedResult.EXPIRED, result);
    }

    @Test
    @DisplayName("옳바른 키 매칭")
    void matchedAllowedKey() {
        UserCredentialsKey.setClock(makeDefaultClock());
        UserCredentialsKey key = new UserCredentialsKey(RAW_CREDENTIALS_KEY);

        UserKeyMatchedResult result = key.matches(RAW_CREDENTIALS_KEY);
        assertEquals(UserKeyMatchedResult.MATCHED, result);
    }
}

File: src/test/kotlin/cube8540/oauth/authentication/user/domain/UserCredentialsKeyTest.kt
package cube8540.oauth.authentication.user.domain

import cube8540.oauth.authentication.AuthenticationApplication
import cube8540.oauth.authentication.toDefaultInstance
import cube8540.oauth.authentication.users.domain.UserCredentialsKey
import cube8540.oauth.authentication.users.domain.UserKeyMatchedResult
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import java.time.Clock
import java.time.LocalDateTime


class UserCredentialsKeyTest {
    private val defaultNow = LocalDateTime.of(2021, 2, 22, 23, 41)
    private val expirationDateTime = defaultNow.plusMinutes(5)

    private val notExpirationNow = expirationDateTime.minusNanos(1)
    private val expirationNow = expirationDateTime.plusNanos(1)

    @Nested
    inner class InitializationTest {
        private val key: UserCredentialsKey

        init {
            UserCredentialsKey.clock = Clock.fixed(defaultNow.toDefaultInstance(), AuthenticationApplication.DEFAULT_TIME_ZONE.toZoneId())
            this.key = UserCredentialsKey("key")
        }

        @Test
        fun `set expiration datetime`() {
            assertThat(key.expiryDateTime).isEqualTo(expirationDateTime)
        }
    }

    @Nested
    inner class MatchTest {
        private val key: UserCredentialsKey

        init {
            UserCredentialsKey.clock = Clock.fixed(defaultNow.toDefaultInstance(), AuthenticationApplication.DEFAULT_TIME_ZONE.toZoneId())
            this.key = UserCredentialsKey("key")
        }

        @Test
        fun `match expired key`() {
            UserCredentialsKey.clock = Clock.fixed(expirationNow.toDefaultInstance(), AuthenticationApplication.DEFAULT_TIME_ZONE.toZoneId())

            val result = key.matches("key")
            assertThat(result).isEqualTo(UserKeyMatchedResult.EXPIRED)
        }

        @Test
        fun `match not equal key`() {
            UserCredentialsKey.clock = Clock.fixed(notExpirationNow.toDefaultInstance(), AuthenticationApplication.DEFAULT_TIME_ZONE.toZoneId())

            val result = key.matches("not matched")
            assertThat(result).isEqualTo(UserKeyMatchedResult.NOT_MATCHED)
        }

        @Test
        fun `match equal key`() {
            UserCredentialsKey.clock = Clock.fixed(notExpirationNow.toDefaultInstance(), AuthenticationApplication.DEFAULT_TIME_ZONE.toZoneId())

            val result = key.matches("key")
            assertThat(result).isEqualTo(UserKeyMatchedResult.MATCHED)
        }
    }
}
